### PR TITLE
Metadata for audio/video/photo files

### PR DESCRIPTION
**Pull request by sergiohs84**
Provides the ability to add coordinates to a file.
Provides the ability to add duration to a file.

### CODE REVIEW

File: src/megaapi_impl.cpp [hunk 1]
```diff
@@ -199,6 +208,23 @@ MegaNodePrivate::MegaNodePrivate(Node *node)
            nameid id = AttrMap::string2nameid(&buf[1]);
            (*customAttrs)[id] = it->second;
        }
+       else
+       {
+           if (buf[0] == 'd')
```
Comment (javiserrano) [inline src/megaapi_impl.cpp]: Check the full name of the attribute

File: src/megaapi_impl.cpp [hunk 2]
```diff
@@ -199,6 +208,23 @@ MegaNodePrivate::MegaNodePrivate(Node *node)
...
+           {
+               memcpy(&duration, it->second.data(), sizeof duration);
```
Comment (javiserrano) [inline src/megaapi_impl.cpp]: Store the attribute value as a (non binary) string

File: src/megaapi_impl.cpp [hunk 3]
```diff
@@ -199,6 +208,23 @@ MegaNodePrivate::MegaNodePrivate(Node *node)
...
+           }
+           else if (it->first == AttrMap::string2nameid("coords"))
+           {
+               string coords = it->second;
+               size_t separator = coords.find_first_of(';');
+               if (separator != coords.npos)
```
Comment (javiserrano) [inline src/megaapi_impl.cpp]: Check if there are enough characters after the separator

File: src/megaapi_impl.cpp [hunk 4]
```diff
@@ -199,6 +208,23 @@ MegaNodePrivate::MegaNodePrivate(Node *node)
...
+               {
+                   latitude = atof(coords.substr(0, separator).c_str());
```
Comment (javiserrano) [inline src/megaapi_impl.cpp]: Use strtod to detect invalid conversions

File: src/megaapi_impl.cpp [hunk 5]
```diff
@@ -4058,6 +4099,38 @@ void MegaApiImpl::setCustomNodeAttribute(MegaNode *node, const char *attrName, c
...
+{
+    MegaRequestPrivate *request = new MegaRequestPrivate(MegaRequest::TYPE_SET_ATTR_NODE, listener);
+    if(node) request->setNodeHandle(node->getHandle());
+    request->setParamType(MegaApi::NODE_ATTR_COORDINATES);
+
+    // calculate the required size of the string containing both values
+    size_t size = 0;
+    std::ostringstream strs;
+    strs << (int) latitude << (int) longitude;
+    size = strs.str().length() + 1 + 6 * 2 + 1;
```
Comment (javiserrano) [inline src/megaapi_impl.cpp]: It should be 7*2 due to the decimal point

File: src/megaapi_impl.cpp [hunk 6]
```diff
@@ -11278,23 +11350,107 @@ void MegaApiImpl::sendPendingRequests()
...
             }
 
-            string sname = attrName;
-            fsAccess->normalize(&sname);
-            sname.insert(0, "_");
-            nameid attr = AttrMap::string2nameid(sname.c_str());
+            if (isOfficial)
+            {
+                int type = request->getParamType();
+                int secs = request->getNumber();
```
Comment (javiserrano) [inline src/megaapi_impl.cpp]: Move this line into the if below

File: src/megaapi_impl.cpp [hunk 7]
```diff
@@ -11278,23 +11350,107 @@ void MegaApiImpl::sendPendingRequests()
...
-            if (attrValue)
+                    if (secs == MegaNode::INVALID_DURATION)
+                    {
+                        node->attrs.map.erase('d');
+                    }
+                    else
+                    {
+                        string buf;
+                        buf.assign((char*)&secs, sizeof secs);
+                        node->attrs.map['d'] = buf;
```
Comment (javiserrano) [inline src/megaapi_impl.cpp]: Save the attribute as a (non binary) string

File: src/megaapi_impl.cpp [hunk 8]
```diff
@@ -11278,23 +11350,107 @@ void MegaApiImpl::sendPendingRequests()
...
+                    {
+                        e = API_EARGS;
+                        break;
+                    }
+
+                    string coordsValue = request->getText();
+                    nameid coordsName = AttrMap::string2nameid("coords");
+
+                    size_t separator = coordsValue.find_first_of(';');
+                    if (separator == coordsValue.npos)
```
Comment (javiserrano) [inline src/megaapi_impl.cpp]: Check if there are enough characters after the separator

File: src/megaapi_impl.cpp [hunk 9]
```diff
@@ -11278,23 +11350,107 @@ void MegaApiImpl::sendPendingRequests()
...
+                    {
+                        e = API_EARGS;
+                        break;
+                    }
+
+                    float latValue = atof(coordsValue.substr(0, separator).c_str());
```
Comment (javiserrano) [inline src/megaapi_impl.cpp]: Use strtod

File: tests/sdk_test.cpp [hunk 10]
```diff
@@ -717,6 +717,116 @@ TEST_F(SdkTest, DISABLED_SdkTestCreateAccount)
...
+    ASSERT_EQ(MegaError::API_OK, lastError[0]) << "Cannot reset node duration (error: " << lastError[0] << ")";
+
+    delete n1;
+    n1 = megaApi[0]->getNodeByHandle(h);
+    ASSERT_EQ(-1, n1->getDuration()) << "Duration value does not match";
+
+
+    // ___ Set invalid coordinates of a node (out of range) ___
+
+    testingInvalidArgs = true;
```
Comment (javiserrano) [inline tests/sdk_test.cpp]: Initialize in SetUp()

Comment (sergiohs84) [inline tests/sdk_test.cpp]: Done!

**Comment by javiserrano** [comment]
Reviewed. All OK!
Ready to be merged when the format of the information is approved.

**Pull request closed without merging**